spring-integration-core/src/main/java/org/springframework/integration/support/json/EmbeddedJsonHeadersMessageMapper.java

```diff
@@ -0,0 +1,255 @@
+		this(JacksonJsonUtils.messagingAwareMapper(), headerPatterns);
+	}
+
+	/**
+	 * Construct an instance that embeds headers matching the supplied patterns using the
+	 * supplied JSON object mapper.
+	 * @param objectMapper the object mapper.
+	 * @param headerPatterns the patterns.
+	 */
+	public EmbeddedJsonHeadersMessageMapper(ObjectMapper objectMapper, Collection<Pattern> headerPatterns) {
+		this.objectMapper = objectMapper;
+		this.headerPatterns = new ArrayList<>(headerPatterns);
+		this.allHeaders = this.headerPatterns.size() == 1 && this.headerPatterns.get(0).pattern().equals(".*");
+	}
+
+	/**
+	 * Set to false to encode byte[] payloads as base64 in JSON. When true,
+	 * byte[] payloads are rendered as discussed in the class description.
+	 * @param rawBytes false to encode as base64.
+	 */
+	public void setRawBytes(boolean rawBytes) {
+		this.rawBytes = rawBytes;
+	}
+
+	public Collection<Pattern> getHeaderPatterns() {
+		return Collections.unmodifiableCollection(this.headerPatterns);
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public byte[] fromMessage(Message<?> message) throws Exception {
+		Message<?> messageToEncode = this.allHeaders ? message : pruneHeaders(message);
+		if (this.rawBytes && message.getPayload() instanceof byte[]) {
+			return fromBytesPayload((Message<byte[]>) messageToEncode);
+		}
+		else {
+			return this.objectMapper.writeValueAsBytes(messageToEncode);
+		}
+	}
+
+	private Message<?> pruneHeaders(Message<?> message) {
+		Map<String, Object> headersToEmbed = new HashMap<>();
+		message.getHeaders().forEach((k, v) -> {
+			if (this.headerPatterns.stream().anyMatch(p -> p.matcher(k).matches())) {
+				headersToEmbed.put(k, v);
+			}
+		});
+		return new MutableMessage<>(message.getPayload(), headersToEmbed);
+	}
+
+	private byte[] fromBytesPayload(Message<byte[]> message) throws UnsupportedEncodingException, Exception {
+		byte[] headers = this.objectMapper.writeValueAsBytes(message.getHeaders());
+		byte[] payload = message.getPayload();
+		ByteBuffer buffer = ByteBuffer.wrap(new byte[8 + headers.length + payload.length]);
+		buffer.putInt(headers.length);
+		buffer.put(headers);
+		buffer.putInt(payload.length);
+		buffer.put(payload);
+		return buffer.array();
+	}
+
+	@Override
+	public Message<?> toMessage(byte[] bytes) throws Exception {
+		Message<?> message = null;
+		try {
+			message = decodeNativeFormat(bytes);
+		}
+		catch (Exception e) {
+			// empty
+		}
+		if (message == null) {
+			try {
+				message = decodeSCStFormat(bytes);
+			}
+			catch (Exception e) {
+				// empty
+			}
+		}
+		if (message == null) {
+			try {
+				message = (Message<?>) this.objectMapper.readValue(bytes, Object.class);
+			}
+			catch (Exception e) {
+				// empty
+			}
+		}
+		if (message != null) {
+			return message;
+		}
+		else {
+			return new GenericMessage<>(bytes);
+		}
+	}
+
+	private Message<?> decodeNativeFormat(byte[] bytes) throws Exception {
```

Comment:
Means that one which belongs to SI per se.
This is SI precious 😄 

Well, in other places we usually use *default*, but Gary decided to go a bit further and just buy a patent to this format I guess 😉 

Seriously I don't know, but I'm fine with this because SI deals with many other formats and protocols, so that it pretty fine to introduce our own and name it **native**  from the SI perspective.

See description in the JavaDocs of this class BTW